### ISSUE TITLE
ST11 - Joined table not referenced elsewhere when using quote

### ISSUE THREAD

**Issue by cedsbe** [sqlfluff/sqlfluff#6719]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

ST11 is raised on the following query:

```sql
SELECT
    [t1].[hello],
    [t2].[world]
FROM
    [table1] AS [t1]
LEFT JOIN [table2] AS [t2] ON [t1].[id] = [t2].[id]
```

While it is not raised on this query:

```sql
SELECT
    t1.[hello],
    t2.[world]
FROM
    [table1] AS t1
LEFT JOIN [table2] AS t2 ON t1.[id] = t2.[id]
```

### Expected Behaviour

The usage of quotes should not impact the rule.

### Observed Behaviour

SQLFluff output:

```
== [repro_fail.sql] FAIL
L:   5 | P:   5 | ST11 | Joined table '[table1] AS [t1]' not referenced elsewhere
                       | in query [structure.unused_join]
L:   6 | P:  11 | ST11 | Joined table '[table2] AS [t2]' not referenced elsewhere
                       | in query [structure.unused_join]
All Finished 📜 🎉!
```

Here is the verbose log for ST11

```
sqlfluff lint "repro_fail.sql" -r ST11 -vvvv
==== sqlfluff ====
sqlfluff:                3.3.1 python:                3.10.12
implementation:        cpython verbosity:                   4
dialect:                  tsql templater:               jinja
rules:                               ST11
== Raw Config:
core:
    dialect:            tsql
    disable_noqa:       False
    encoding:           autodetect
    fix_even_unparsable:False
    ignore:             []
    ignore_templated_areas:True
    large_file_skip_byte_limit:20000
    large_file_skip_char_limit:0
    max_line_length:    80
    nocolor:            False
    output_line_length: 80
    processes:          1
    render_variant_limit:1
    require_dialect:    False
    rules:              ST11
    runaway_limit:      10
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          jinja
    verbose:            4
    warn_unused_ignores:False
    warnings:           []
indentation:
    allow_implicit_indents:False
    ignore_comment_lines:False
    indent_unit:        space
    indented_ctes:      False
    indented_joins:     False
    indented_on_contents:True
    indented_then:      True
    indented_then_contents:True
    indented_using_on:  True
    skip_indentation_in:script_content
    tab_space_size:     4
    template_blocks_indent:True
    trailing_comments:  before
layout:
    type:
        array_accessor:
            spacing_before:     touch:inline
        array_type:
            spacing_within:     touch:inline
        assignment_operator:
            line_position:      leading
            spacing_within:     touch
        binary_operator:
            line_position:      leading
            spacing_within:     touch
        bracketed_arguments:
            spacing_before:     touch:inline
        casting_operator:
            spacing_after:      touch:inline
            spacing_before:     touch
        colon:
            spacing_before:     touch
        colon_delimiter:
            spacing_after:      touch
            spacing_before:     touch
        column_path_operator:
            line_position:      leading
            spacing_within:     touch
        comma:
            line_position:      trailing
            spacing_before:     touch
        comment:
            spacing_after:      any
            spacing_before:     any
        common_table_expression:
            spacing_within:     single:inline
        comparison_operator:
            line_position:      leading
            spacing_within:     touch
        dot:
            spacing_after:      touch
            spacing_before:     touch
        end_angle_bracket:
            spacing_before:     touch
        end_bracket:
            spacing_before:     touch
        end_of_file:
            spacing_before:     touch
        end_square_bracket:
            spacing_before:     touch
        from_clause:
            line_position:      alone
        function_contents:
            spacing_before:     touch:inline
        function_name:
            spacing_within:     touch:inline
        function_parameter_list:
            spacing_before:     touch:inline
        groupby_clause:
            keyword_line_position:leading
            line_position:      alone
        having_clause:
            keyword_line_position:leading
            line_position:      alone
        join_clause:
            line_position:      alone
        limit_clause:
            line_position:      alone
        match_condition:
            spacing_within:     touch:inline
        numeric_literal:
            spacing_within:     touch:inline
        object_reference:
            spacing_within:     touch:inline
        orderby_clause:
            keyword_line_position:leading
            line_position:      leading
        path_segment:
            spacing_within:     touch
        pattern_expression:
            spacing_within:     any
        placeholder:
            spacing_after:      any
            spacing_before:     any
        select_clause:
            line_position:      alone
        semi_structured_expression:
            spacing_before:     touch:inline
            spacing_within:     touch:inline
        set_operator:
            line_position:      alone:strict
        sign_indicator:
            spacing_after:      touch:inline
        sized_array_type:
            spacing_within:     touch
        slash:
            spacing_after:      any
            spacing_before:     any
        slice:
            spacing_after:      touch
            spacing_before:     touch
        sql_conf_option:
            spacing_within:     touch
        sqlcmd_operator:
            spacing_before:     touch
        start_angle_bracket:
            spacing_after:      touch
        start_bracket:
            spacing_after:      touch
        start_square_bracket:
            spacing_after:      touch
        statement_terminator:
            line_position:      trailing
            spacing_before:     touch
        struct_type:
            spacing_within:     touch:inline
        template_loop:
            spacing_after:      any
            spacing_before:     any
        tilde:
            spacing_after:      touch:inline
        typed_array_literal:
            spacing_within:     touch
        typed_struct_literal:
            spacing_within:     touch
        where_clause:
            keyword_line_position:leading
            line_position:      alone
rules:
    allow_scalar:       True
    single_table_references:consistent
    unquoted_identifiers_policy:all
    aliasing.column:
        aliasing:           explicit
    aliasing.forbid:
        force_enable:       False
    aliasing.length:
    aliasing.table:
        aliasing:           explicit
    aliasing.unused:
        alias_case_check:   dialect
    ambiguous.column_references:
        group_by_and_order_by_style:consistent
    ambiguous.join:
        fully_qualify_join_types:inner
    capitalisation.functions:
        extended_capitalisation_policy:consistent
    capitalisation.identifiers:
        extended_capitalisation_policy:consistent
    capitalisation.keywords:
        capitalisation_policy:consistent
    capitalisation.literals:
        capitalisation_policy:consistent
    capitalisation.types:
        extended_capitalisation_policy:consistent
    convention.blocked_words:
        match_source:       False
    convention.casting_style:
        preferred_type_casting_style:consistent
    convention.count_rows:
        prefer_count_0:     False
        prefer_count_1:     False
    convention.not_equal:
        preferred_not_equal_style:consistent
    convention.quoted_literals:
        force_enable:       False
        preferred_quoted_literal_style:consistent
    convention.select_trailing_comma:
        select_clause_trailing_comma:forbid
    convention.terminator:
        multiline_newline:  False
        require_final_semicolon:False
    layout.long_lines:
        ignore_comment_clauses:False
        ignore_comment_lines:False
    layout.select_targets:
        wildcard_policy:    single
    references.consistent:
        force_enable:       False
    references.from:
        force_enable:       False
    references.keywords:
        unquoted_identifiers_policy:aliases
    references.qualification:
    references.quoting:
        case_sensitive:     True
        prefer_quoted_identifiers:True
        prefer_quoted_keywords:False
    references.special_chars:
        allow_space_in_identifier:False
        quoted_identifiers_policy:all
        unquoted_identifiers_policy:all
    structure.join_condition_order:
        preferred_first_table_in_join_clause:earlier
    structure.subquery:
        forbid_subquery_in: join
templater:
    unwrap_wrapped_queries:True
    jinja:
        apply_dbt_builtins: True

==== readout ====

INFO       Rendering String [jinja] (repro_fail.sql)
INFO       Slicing File Template
DEBUG          Raw String: 'SELECT\n    [t1].[hello],\n    [t2].[world]\nFROM\n    [table1] AS [t1]\nLEFT JOIN [t'
INFO       Rendered 1 variants
INFO       Parse Rendered. Lexing Variant 0
INFO       LEXING RAW (repro_fail.sql)
INFO       Elements to Segments.
DEBUG        0: TemplateElement(raw='SELECT', template_slice=slice(0, 6, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        1: TemplateElement(raw='\n', template_slice=slice(6, 7, None), matcher=<RegexLexer: newline>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        2: TemplateElement(raw='    ', template_slice=slice(7, 11, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        3: TemplateElement(raw='[t1]', template_slice=slice(11, 15, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        4: TemplateElement(raw='.', template_slice=slice(15, 16, None), matcher=<StringLexer: dot>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        5: TemplateElement(raw='[hello]', template_slice=slice(16, 23, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        6: TemplateElement(raw=',', template_slice=slice(23, 24, None), matcher=<StringLexer: comma>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        7: TemplateElement(raw='\n', template_slice=slice(24, 25, None), matcher=<RegexLexer: newline>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        8: TemplateElement(raw='    ', template_slice=slice(25, 29, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        9: TemplateElement(raw='[t2]', template_slice=slice(29, 33, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        10: TemplateElement(raw='.', template_slice=slice(33, 34, None), matcher=<StringLexer: dot>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        11: TemplateElement(raw='[world]', template_slice=slice(34, 41, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        12: TemplateElement(raw='\n', template_slice=slice(41, 42, None), matcher=<RegexLexer: newline>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        13: TemplateElement(raw='FROM', template_slice=slice(42, 46, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        14: TemplateElement(raw='\n', template_slice=slice(46, 47, None), matcher=<RegexLexer: newline>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        15: TemplateElement(raw='    ', template_slice=slice(47, 51, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        16: TemplateElement(raw='[table1]', template_slice=slice(51, 59, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        17: TemplateElement(raw=' ', template_slice=slice(59, 60, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        18: TemplateElement(raw='AS', template_slice=slice(60, 62, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        19: TemplateElement(raw=' ', template_slice=slice(62, 63, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        20: TemplateElement(raw='[t1]', template_slice=slice(63, 67, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        21: TemplateElement(raw='\n', template_slice=slice(67, 68, None), matcher=<RegexLexer: newline>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        22: TemplateElement(raw='LEFT', template_slice=slice(68, 72, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        23: TemplateElement(raw=' ', template_slice=slice(72, 73, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        24: TemplateElement(raw='JOIN', template_slice=slice(73, 77, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        25: TemplateElement(raw=' ', template_slice=slice(77, 78, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        26: TemplateElement(raw='[table2]', template_slice=slice(78, 86, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        27: TemplateElement(raw=' ', template_slice=slice(86, 87, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        28: TemplateElement(raw='AS', template_slice=slice(87, 89, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        29: TemplateElement(raw=' ', template_slice=slice(89, 90, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        30: TemplateElement(raw='[t2]', template_slice=slice(90, 94, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        31: TemplateElement(raw=' ', template_slice=slice(94, 95, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        32: TemplateElement(raw='ON', template_slice=slice(95, 97, None), matcher=<RegexLexer: word>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        33: TemplateElement(raw=' ', template_slice=slice(97, 98, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        34: TemplateElement(raw='[t1]', template_slice=slice(98, 102, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        35: TemplateElement(raw='.', template_slice=slice(102, 103, None), matcher=<StringLexer: dot>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        36: TemplateElement(raw='[id]', template_slice=slice(103, 107, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        37: TemplateElement(raw=' ', template_slice=slice(107, 108, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        38: TemplateElement(raw='=', template_slice=slice(108, 109, None), matcher=<StringLexer: equals>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        39: TemplateElement(raw=' ', template_slice=slice(109, 110, None), matcher=<RegexLexer: whitespace>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        40: TemplateElement(raw='[t2]', template_slice=slice(110, 114, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        41: TemplateElement(raw='.', template_slice=slice(114, 115, None), matcher=<StringLexer: dot>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        42: TemplateElement(raw='[id]', template_slice=slice(115, 119, None), matcher=<RegexLexer: square_quote>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
DEBUG        43: TemplateElement(raw='\n', template_slice=slice(119, 120, None), matcher=<RegexLexer: newline>). [tfs_idx = 0]
DEBUG            0: TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 120, None), templated_slice=slice(0, 120, None))
DEBUG           Consuming whole from literal. Existing Consumed: 0
INFO       Lexed segments: ['SELECT', '\n', '    ', '[t1]', '.', '[hello]', ',', '\n', '    ', '[t2]', '.', '[world]', '\n', 'FROM', '\n', '    ', '[table1]', ' ', 'AS', ' ', '[t1]', '\n', 'LEFT', ' ', 'JOIN', ' ', '[table2]', ' ', 'AS', ' ', '[t2]', ' ', 'ON', ' ', '[t1]', '.', '[id]', ' ', '=', ' ', '[t2]', '.', '[id]', '\n', '']
INFO       Parse Rendered. Parsing Variant 0
INFO       Root Match:
Match (None): slice(0, 43, None)
  +Match (<class 'sqlfluff.dialects.dialect_tsql.BatchSegment'>): slice(0, 43, None)
    +Match (<class 'sqlfluff.dialects.dialect_tsql.StatementSegment'>): slice(0, 43, None)
      +Match (<class 'sqlfluff.dialects.dialect_tsql.SelectStatementSegment'>): slice(0, 43, None)
        +Match (<class 'sqlfluff.dialects.dialect_tsql.SelectClauseSegment'>): slice(0, 12, None)
          +1: <class 'sqlfluff.core.parser.segments.meta.Indent'>
          +12: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
          +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(0, 1, None)
            -instance_types: ('keyword',)
          +Match (<class 'sqlfluff.dialects.dialect_tsql.SelectClauseElementSegment'>): slice(3, 6, None)
            +Match (<class 'sqlfluff.dialects.dialect_tsql.ColumnReferenceSegment'>): slice(3, 6, None)
              +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(3, 4, None)
                -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                -casefold: <method 'upper' of 'str' objects>
              +Match (<class 'sqlfluff.core.parser.segments.common.SymbolSegment'>): slice(4, 5, None)
                -instance_types: ('dot',)
              +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(5, 6, None)
                -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                -casefold: <method 'upper' of 'str' objects>
          +Match (<class 'sqlfluff.core.parser.segments.common.SymbolSegment'>): slice(6, 7, None)
            -instance_types: ('comma',)
          +Match (<class 'sqlfluff.dialects.dialect_tsql.SelectClauseElementSegment'>): slice(9, 12, None)
            +Match (<class 'sqlfluff.dialects.dialect_tsql.ColumnReferenceSegment'>): slice(9, 12, None)
              +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(9, 10, None)
                -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                -casefold: <method 'upper' of 'str' objects>
              +Match (<class 'sqlfluff.core.parser.segments.common.SymbolSegment'>): slice(10, 11, None)
                -instance_types: ('dot',)
              +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(11, 12, None)
                -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                -casefold: <method 'upper' of 'str' objects>
        +Match (<class 'sqlfluff.dialects.dialect_tsql.FromClauseSegment'>): slice(13, 43, None)
          +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(13, 14, None)
            -instance_types: ('keyword',)
          +Match (<class 'sqlfluff.dialects.dialect_ansi.FromExpressionSegment'>): slice(16, 43, None)
            +16: <class 'sqlfluff.core.parser.segments.meta.Indent'>
            +22: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
            +Match (<class 'sqlfluff.dialects.dialect_ansi.FromExpressionElementSegment'>): slice(16, 21, None)
              +Match (<class 'sqlfluff.dialects.dialect_tsql.TableExpressionSegment'>): slice(16, 17, None)
                +Match (<class 'sqlfluff.dialects.dialect_tsql.TableReferenceSegment'>): slice(16, 17, None)
                  +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(16, 17, None)
                    -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                    -casefold: <method 'upper' of 'str' objects>
              +Match (<class 'sqlfluff.dialects.dialect_ansi.AliasExpressionSegment'>): slice(18, 21, None)
                +18: <class 'sqlfluff.core.parser.segments.meta.Indent'>
                +21: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
                +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(18, 19, None)
                  -instance_types: ('keyword',)
                +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(20, 21, None)
                  -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                  -casefold: <method 'upper' of 'str' objects>
            +Match (<class 'sqlfluff.dialects.dialect_ansi.JoinClauseSegment'>): slice(22, 43, None)
              +25: <class 'sqlfluff.core.parser.segments.meta.Indent'>
              +32: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
              +32: <class 'sqlfluff.core.parser.segments.meta.Indent'>
              +43: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
              +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(22, 23, None)
                -instance_types: ('keyword',)
              +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(24, 25, None)
                -instance_types: ('keyword',)
              +Match (<class 'sqlfluff.dialects.dialect_ansi.FromExpressionElementSegment'>): slice(26, 31, None)
                +Match (<class 'sqlfluff.dialects.dialect_tsql.TableExpressionSegment'>): slice(26, 27, None)
                  +Match (<class 'sqlfluff.dialects.dialect_tsql.TableReferenceSegment'>): slice(26, 27, None)
                    +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(26, 27, None)
                      -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                      -casefold: <method 'upper' of 'str' objects>
                +Match (<class 'sqlfluff.dialects.dialect_ansi.AliasExpressionSegment'>): slice(28, 31, None)
                  +28: <class 'sqlfluff.core.parser.segments.meta.Indent'>
                  +31: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
                  +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(28, 29, None)
                    -instance_types: ('keyword',)
                  +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(30, 31, None)
                    -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                    -casefold: <method 'upper' of 'str' objects>
              +Match (<class 'sqlfluff.dialects.dialect_ansi.JoinOnConditionSegment'>): slice(32, 43, None)
                +33: <class 'sqlfluff.core.parser.segments.meta.ImplicitIndent'>
                +43: <class 'sqlfluff.core.parser.segments.meta.Dedent'>
                +Match (<class 'sqlfluff.core.parser.segments.keyword.KeywordSegment'>): slice(32, 33, None)
                  -instance_types: ('keyword',)
                +Match (<class 'sqlfluff.dialects.dialect_tsql.ExpressionSegment'>): slice(34, 43, None)
                  +Match (<class 'sqlfluff.dialects.dialect_tsql.ColumnReferenceSegment'>): slice(34, 37, None)
                    +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(34, 35, None)
                      -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                      -casefold: <method 'upper' of 'str' objects>
                    +Match (<class 'sqlfluff.core.parser.segments.common.SymbolSegment'>): slice(35, 36, None)
                      -instance_types: ('dot',)
                    +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(36, 37, None)
                      -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                      -casefold: <method 'upper' of 'str' objects>
                  +Match (<class 'sqlfluff.dialects.dialect_ansi.EqualsSegment'>): slice(38, 39, None)
                    +Match (<class 'sqlfluff.core.parser.segments.common.SymbolSegment'>): slice(38, 39, None)
                      -instance_types: ('raw_comparison_operator',)
                  +Match (<class 'sqlfluff.dialects.dialect_tsql.ColumnReferenceSegment'>): slice(40, 43, None)
                    +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(40, 41, None)
                      -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                      -casefold: <method 'upper' of 'str' objects>
                    +Match (<class 'sqlfluff.core.parser.segments.common.SymbolSegment'>): slice(41, 42, None)
                      -instance_types: ('dot',)
                    +Match (<class 'sqlfluff.core.parser.segments.common.IdentifierSegment'>): slice(42, 43, None)
                      -instance_types: ('quoted_identifier', 'identifier', 'square_quote')
                      -casefold: <method 'upper' of 'str' objects>
INFO
###
#
# Parsed Tree:
#
###
INFO
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    batch:
[L:  1, P:  1]      |        statement:
[L:  1, P:  1]      |            select_statement:
[L:  1, P:  1]      |                select_clause:
[L:  1, P:  1]      |                    keyword:                                  'SELECT'
[L:  1, P:  7]      |                    [META] indent:
[L:  1, P:  7]      |                    newline:                                  '\n'
[L:  2, P:  1]      |                    whitespace:                               '    '
[L:  2, P:  5]      |                    select_clause_element:
[L:  2, P:  5]      |                        column_reference:
[L:  2, P:  5]      |                            quoted_identifier:                '[t1]'
[L:  2, P:  9]      |                            dot:                              '.'
[L:  2, P: 10]      |                            quoted_identifier:                '[hello]'
[L:  2, P: 17]      |                    comma:                                    ','
[L:  2, P: 18]      |                    newline:                                  '\n'
[L:  3, P:  1]      |                    whitespace:                               '    '
[L:  3, P:  5]      |                    select_clause_element:
[L:  3, P:  5]      |                        column_reference:
[L:  3, P:  5]      |                            quoted_identifier:                '[t2]'
[L:  3, P:  9]      |                            dot:                              '.'
[L:  3, P: 10]      |                            quoted_identifier:                '[world]'
[L:  3, P: 17]      |                    [META] dedent:
[L:  3, P: 17]      |                newline:                                      '\n'
[L:  4, P:  1]      |                from_clause:
[L:  4, P:  1]      |                    keyword:                                  'FROM'
[L:  4, P:  5]      |                    newline:                                  '\n'
[L:  5, P:  1]      |                    whitespace:                               '    '
[L:  5, P:  5]      |                    from_expression:
[L:  5, P:  5]      |                        [META] indent:
[L:  5, P:  5]      |                        from_expression_element:
[L:  5, P:  5]      |                            table_expression:
[L:  5, P:  5]      |                                table_reference:
[L:  5, P:  5]      |                                    quoted_identifier:        '[table1]'
[L:  5, P: 13]      |                            whitespace:                       ' '
[L:  5, P: 14]      |                            alias_expression:
[L:  5, P: 14]      |                                [META] indent:
[L:  5, P: 14]      |                                keyword:                      'AS'
[L:  5, P: 16]      |                                whitespace:                   ' '
[L:  5, P: 17]      |                                quoted_identifier:            '[t1]'
[L:  5, P: 21]      |                                [META] dedent:
[L:  5, P: 21]      |                        newline:                              '\n'
[L:  6, P:  1]      |                        [META] dedent:
[L:  6, P:  1]      |                        join_clause:
[L:  6, P:  1]      |                            keyword:                          'LEFT'
[L:  6, P:  5]      |                            whitespace:                       ' '
[L:  6, P:  6]      |                            keyword:                          'JOIN'
[L:  6, P: 10]      |                            [META] indent:
[L:  6, P: 10]      |                            whitespace:                       ' '
[L:  6, P: 11]      |                            from_expression_element:
[L:  6, P: 11]      |                                table_expression:
[L:  6, P: 11]      |                                    table_reference:
[L:  6, P: 11]      |                                        quoted_identifier:    '[table2]'
[L:  6, P: 19]      |                                whitespace:                   ' '
[L:  6, P: 20]      |                                alias_expression:
[L:  6, P: 20]      |                                    [META] indent:
[L:  6, P: 20]      |                                    keyword:                  'AS'
[L:  6, P: 22]      |                                    whitespace:               ' '
[L:  6, P: 23]      |                                    quoted_identifier:        '[t2]'
[L:  6, P: 27]      |                                    [META] dedent:
[L:  6, P: 27]      |                            whitespace:                       ' '
[L:  6, P: 28]      |                            [META] dedent:
[L:  6, P: 28]      |                            [META] indent:
[L:  6, P: 28]      |                            join_on_condition:
[L:  6, P: 28]      |                                keyword:                      'ON'
[L:  6, P: 30]      |                                [META] (implicit) indent:
[L:  6, P: 30]      |                                whitespace:                   ' '
[L:  6, P: 31]      |                                expression:
[L:  6, P: 31]      |                                    column_reference:
[L:  6, P: 31]      |                                        quoted_identifier:    '[t1]'
[L:  6, P: 35]      |                                        dot:                  '.'
[L:  6, P: 36]      |                                        quoted_identifier:    '[id]'
[L:  6, P: 40]      |                                    whitespace:               ' '
[L:  6, P: 41]      |                                    comparison_operator:
[L:  6, P: 41]      |                                        raw_comparison_operator:  '='
[L:  6, P: 42]      |                                    whitespace:               ' '
[L:  6, P: 43]      |                                    column_reference:
[L:  6, P: 43]      |                                        quoted_identifier:    '[t2]'
[L:  6, P: 47]      |                                        dot:                  '.'
[L:  6, P: 48]      |                                        quoted_identifier:    '[id]'
[L:  6, P: 52]      |                                [META] dedent:
[L:  6, P: 52]      |                            [META] dedent:
[L:  6, P: 52]      |    newline:                                                  '\n'
[L:  7, P:  1]      |    [META] end_of_file:

INFO       Parse Rendered. Variant 0. Lex in 0.007148387001507217. Parse in 0.014995757999713533.
INFO       lint_parsed - linting root variant (repro_fail.sql)
== [repro_fail.sql] LINTING (ST11)
INFO

Entering linter phase main, loop 1/1

DEBUG      [ST11] Processed SELECT statement.
Joined tables in scope: [('[T1]', <FromExpressionElementSegment: ([L:  5, P:  5])>), ('[T2]', <FromExpressionElementSegment: ([L:  6, P: 11])>)]
...of which referenced in non-self join clauses: ['T1', 'T2']
DEBUG      [ST11]     'T1' referenced in <SelectClauseSegment: ([L:  1, P:  1])>
DEBUG      [ST11]     'T2' referenced in <SelectClauseSegment: ([L:  1, P:  1])>
DEBUG      [ST11] Select statement <SelectStatementSegment: ([L:  1, P:  1])> references tables: {'T2', 'T1'}.
Joined tables to asses: [('[T1]', <FromExpressionElementSegment: ([L:  5, P:  5])>), ('[T2]', <FromExpressionElementSegment: ([L:  6, P: 11])>)]
INFO       [ST11] !! Violation Found: "Joined table '[table1] AS [t1]' not referenced elsewhere in query"
INFO       [ST11] !! Violation Found: "Joined table '[table2] AS [t2]' not referenced elsewhere in query"
INFO
###
#
# Fixed Tree:
#
###
INFO
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    batch:
[L:  1, P:  1]      |        statement:
[L:  1, P:  1]      |            select_statement:
[L:  1, P:  1]      |                select_clause:
[L:  1, P:  1]      |                    keyword:                                  'SELECT'
[L:  1, P:  7]      |                    [META] indent:
[L:  1, P:  7]      |                    newline:                                  '\n'
[L:  2, P:  1]      |                    whitespace:                               '    '
[L:  2, P:  5]      |                    select_clause_element:
[L:  2, P:  5]      |                        column_reference:
[L:  2, P:  5]      |                            quoted_identifier:                '[t1]'
[L:  2, P:  9]      |                            dot:                              '.'
[L:  2, P: 10]      |                            quoted_identifier:                '[hello]'
[L:  2, P: 17]      |                    comma:                                    ','
[L:  2, P: 18]      |                    newline:                                  '\n'
[L:  3, P:  1]      |                    whitespace:                               '    '
[L:  3, P:  5]      |                    select_clause_element:
[L:  3, P:  5]      |                        column_reference:
[L:  3, P:  5]      |                            quoted_identifier:                '[t2]'
[L:  3, P:  9]      |                            dot:                              '.'
[L:  3, P: 10]      |                            quoted_identifier:                '[world]'
[L:  3, P: 17]      |                    [META] dedent:
[L:  3, P: 17]      |                newline:                                      '\n'
[L:  4, P:  1]      |                from_clause:
[L:  4, P:  1]      |                    keyword:                                  'FROM'
[L:  4, P:  5]      |                    newline:                                  '\n'
[L:  5, P:  1]      |                    whitespace:                               '    '
[L:  5, P:  5]      |                    from_expression:
[L:  5, P:  5]      |                        [META] indent:
[L:  5, P:  5]      |                        from_expression_element:
[L:  5, P:  5]      |                            table_expression:
[L:  5, P:  5]      |                                table_reference:
[L:  5, P:  5]      |                                    quoted_identifier:        '[table1]'
[L:  5, P: 13]      |                            whitespace:                       ' '
[L:  5, P: 14]      |                            alias_expression:
[L:  5, P: 14]      |                                [META] indent:
[L:  5, P: 14]      |                                keyword:                      'AS'
[L:  5, P: 16]      |                                whitespace:                   ' '
[L:  5, P: 17]      |                                quoted_identifier:            '[t1]'
[L:  5, P: 21]      |                                [META] dedent:
[L:  5, P: 21]      |                        newline:                              '\n'
[L:  6, P:  1]      |                        [META] dedent:
[L:  6, P:  1]      |                        join_clause:
[L:  6, P:  1]      |                            keyword:                          'LEFT'
[L:  6, P:  5]      |                            whitespace:                       ' '
[L:  6, P:  6]      |                            keyword:                          'JOIN'
[L:  6, P: 10]      |                            [META] indent:
[L:  6, P: 10]      |                            whitespace:                       ' '
[L:  6, P: 11]      |                            from_expression_element:
[L:  6, P: 11]      |                                table_expression:
[L:  6, P: 11]      |                                    table_reference:
[L:  6, P: 11]      |                                        quoted_identifier:    '[table2]'
[L:  6, P: 19]      |                                whitespace:                   ' '
[L:  6, P: 20]      |                                alias_expression:
[L:  6, P: 20]      |                                    [META] indent:
[L:  6, P: 20]      |                                    keyword:                  'AS'
[L:  6, P: 22]      |                                    whitespace:               ' '
[L:  6, P: 23]      |                                    quoted_identifier:        '[t2]'
[L:  6, P: 27]      |                                    [META] dedent:
[L:  6, P: 27]      |                            whitespace:                       ' '
[L:  6, P: 28]      |                            [META] dedent:
[L:  6, P: 28]      |                            [META] indent:
[L:  6, P: 28]      |                            join_on_condition:
[L:  6, P: 28]      |                                keyword:                      'ON'
[L:  6, P: 30]      |                                [META] (implicit) indent:
[L:  6, P: 30]      |                                whitespace:                   ' '
[L:  6, P: 31]      |                                expression:
[L:  6, P: 31]      |                                    column_reference:
[L:  6, P: 31]      |                                        quoted_identifier:    '[t1]'
[L:  6, P: 35]      |                                        dot:                  '.'
[L:  6, P: 36]      |                                        quoted_identifier:    '[id]'
[L:  6, P: 40]      |                                    whitespace:               ' '
[L:  6, P: 41]      |                                    comparison_operator:
[L:  6, P: 41]      |                                        raw_comparison_operator:  '='
[L:  6, P: 42]      |                                    whitespace:               ' '
[L:  6, P: 43]      |                                    column_reference:
[L:  6, P: 43]      |                                        quoted_identifier:    '[t2]'
[L:  6, P: 47]      |                                        dot:                  '.'
[L:  6, P: 48]      |                                        quoted_identifier:    '[id]'
[L:  6, P: 52]      |                                [META] dedent:
[L:  6, P: 52]      |                            [META] dedent:
[L:  6, P: 52]      |    newline:                                                  '\n'
[L:  7, P:  1]      |    [META] end_of_file:

== [repro_fail.sql] FAIL
L:   5 | P:   5 | ST11 | Joined table '[table1] AS [t1]' not referenced elsewhere
                       | in query [structure.unused_join]
L:   6 | P:  11 | ST11 | Joined table '[table2] AS [t2]' not referenced elsewhere
                       | in query [structure.unused_join]
==== summary ====
files:             1 violations:        2
clean files:       0 unclean files:     1
avg per file:   2.00 unclean rate:   100%
status:         FAIL
All Finished 📜 🎉!
```

### How to reproduce

```bash
sqlfluff lint "repro_fail.sql" -r ST11 --dialect tsql
```

Where `repro_fail.sql` is 

```sql
SELECT
    [t1].[hello],
    [t2].[world]
FROM
    [table1] AS [t1]
LEFT JOIN [table2] AS [t2] ON [t1].[id] = [t2].[id]
```

### Dialect

tsql

### Version

sqlfluff:  3.3.1
python: 3.10.12

### Configuration

```ini
[sqlfluff]
dialect = tsql

[sqlfluff:rules:references.quoting]
prefer_quoted_identifiers = True
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by keraion** [comment]
Fixed by #6702
